### PR TITLE
WD-8309 - feat: Display error when trying to retrieve models info

### PR DESCRIPTION
**Pull request by vladimir-cucu**
## Done

- Throw the “Unable to fetch the status.” error instead of using return. Afterward, throw the error, so that it is further caught in `fetchAllModelStatuses()`.
- If less than 10% of models error out, no error is processed in `fetchAllModelStatuses()`. Else, throw the error back, so that it is further caught in `modelPollerMiddleware()`.
- When trying to retrieve models info in `modelPollerMiddleware()`, send "Unable to load models." error if it is the first polling and all models error out, "Unable to load some models." error if it is the first polling and some models error out, "Unable to load latest model data." if it is later polling and the error appears in `fetchAllModelStatuses()` call, else send "Unable to list or update models." error. The error should be sent to `juju -> modelsErrors` in Redux store.
- Display the previously mentioned error inline in `pages/ModelsIndex/ModelsIndex.tsx`.

## QA

- Modify the code of `modelPollerMiddleware()` to throw an error on line 217.
- Verify that the error gets displayed on `Models` page and logged into the console.

## Details

- https://warthogs.atlassian.net/browse/WD-8309

## Screenshots

Screenshot from 2024-01-23 19-39-26

### CODE REVIEW

File: src/store/middleware/model-poller.ts [hunk 1]
```diff
@@ -207,11 +208,34 @@ export const modelPollerMiddleware: Middleware<
...
+                    wsControllerURL,
+                  }),
+                );
+              }
             } catch (e) {
-              console.log(e);
+              console.log("something");
+              const errorMessage = isFirstTimePolling
+                ? "Unable to fetch the models."
+                : "Unable to update the models.";
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/store/middleware/model-poller.ts:225]: Done!

File: src/store/middleware/model-poller.ts [hunk 2]
```diff
@@ -207,11 +208,34 @@ export const modelPollerMiddleware: Middleware<
                 reduxStore.dispatch,
                 reduxStore.getState,
               );
+              // If the code execution arrives here, then the model statuses
+              // have been succesfully updated. Models error should be removed.
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/store/middleware/model-poller.ts:212]: Oops, done :)

File: src/store/middleware/model-poller.ts [hunk 3]
```diff
@@ -186,6 +186,7 @@ export const modelPollerMiddleware: Middleware<
         }
 
         let pollCount = 0;
+        let isFirstTimePolling = true;
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/store/middleware/model-poller.ts:189]: Changed!

File: src/store/middleware/model-poller.ts [hunk 4]
```diff
@@ -207,11 +208,34 @@ export const modelPollerMiddleware: Middleware<
...
+                reduxStore.dispatch(
+                  jujuActions.updateModelsError({
+                    modelsError: null,
+                    wsControllerURL,
+                  }),
+                );
+              }
             } catch (e) {
-              console.log(e);
+              console.log("something");
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/store/middleware/model-poller.ts]: Oops, removed it :)

File: src/juju/api.ts [hunk 5]
```diff
@@ -344,49 +344,60 @@ export async function fetchAllModelStatuses(
...
         }
       }
       done();
     });
   });
-  return new Promise<void>((resolve) => {
+  // If errors appear in more than a certain number of models, a rejected
+  // promise should be returned and further handled in modelPollerMiddleware().
+  const shouldDisplayError =
+    modelErrorCount >= Math.max(5, 0.1 * modelUUIDList.length);
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/juju/api.ts]: Fixed it! Thanks for the review!

File: src/store/juju/slice.ts [hunk 6]
```diff
@@ -128,9 +129,20 @@ const slice = createSlice({
...
     },
+    updateModelsError: (
+      state,
+      action: PayloadAction<
+        {
+          modelsError: string | null;
+        } & WsControllerURLParam
+      >,
+    ) => {
+      state.modelsError = action.payload.modelsError;
```
Comment (vladimir-cucu) [inline src/store/juju/slice.ts:140]: I've added a test for `updateModelsError` in `actions.test.ts`.

File: src/store/juju/slice.ts [hunk 7]
```diff
@@ -128,9 +129,20 @@ const slice = createSlice({
         state.modelData[modelInfo.uuid].info = modelInfo;
       }
     },
+    updateModelsError: (
```
Comment (huwshimi) [inline src/store/juju/slice.ts:132]: It looks like this needs a reducer test.

**Review by huwshimi — APPROVED** [review]
This looks good, one test to add (feel free to merge this once you add that test).

File: src/store/juju/slice.ts (hunk 7 above)
Comment (vladimir-cucu) [reply src/store/juju/slice.ts:132]: Nice catch! Added the test.

**Pull request merged**